src/lib/io/sol-uart-impl-linux.c

```diff
@@ -32,49 +33,147 @@
+
 struct sol_uart {
+    struct sol_fd *fd_handler;
+    const void *user_data;
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, const struct sol_buffer *buf);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    struct sol_timeout *read_timeout;
+    struct sol_ptr_vector pending_blobs;
+    struct sol_buffer rx;
+    size_t tx_size;
+    size_t pending_tx;
+    size_t written; //How many bytes We've written so far for the first element of pending_blobs
     int fd;
-    struct {
-        void *rx_fd_handler;
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void *tx_fd_handler;
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length, tx_index;
-    } async;
+    bool delete_me;
+    bool in_use;
 };
 
 static bool
-uart_rx_callback(void *data, int fd, uint32_t active_flags)
+read_timeout(void *data)
 {
     struct sol_uart *uart = data;
+    ssize_t r;
+    bool keep_running = true;
+
+    uart->in_use = true;
+    r = uart->rx_cb((void *)uart->user_data, uart, &uart->rx);
+    uart->in_use = false;
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    if ((size_t)r < uart->rx.used) {
+        memmove(uart->rx.data, (char *)uart->rx.data + r, uart->rx.used - r);
+        uart->rx.used -= r;
+    } else if ((size_t)r >= uart->rx.used)
+        uart->rx.used = 0;
+
+    if (!uart->rx.used) {
+        keep_running = false;
+        uart->read_timeout = NULL;
+    }
 
-    if (!uart->async.rx_cb)
-        return true;
+exit:
+    if (uart->delete_me) {
+        keep_running = false;
+        uart->read_timeout = NULL;
+        sol_uart_close(uart);
+    }
+    return keep_running;
+}
+
+static bool
+uart_fd_handler_callback(void *data, int fd, uint32_t active_flags)
+{
+    struct sol_uart *uart = data;
+    struct sol_blob *blob;
+    int status = 0;
+    ssize_t r;
 
     if (active_flags & FD_ERROR_FLAGS) {
         SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
         return true;
     }
-    if (active_flags & SOL_FD_FLAGS_IN) {
-        uint8_t c; /* Backing storage for next sol_buffer */
-        struct sol_buffer buf = SOL_BUFFER_INIT_FLAGS(&c, sizeof(c),
-            SOL_BUFFER_FLAGS_MEMORY_NOT_OWNED | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
-        int status = sol_util_fill_buffer(uart->fd, &buf, 1);
-        if (status > 0)
-            uart->async.rx_cb((void *)uart->async.rx_user_data, uart, c);
+
+    if ((active_flags & SOL_FD_FLAGS_IN) && uart->rx_cb) {
+        size_t remaining = uart->rx.capacity - uart->rx.used;
+
+        if (!remaining && !(uart->rx.flags & SOL_BUFFER_FLAGS_FIXED_CAPACITY)) {
+            int err;
+
+            err = sol_buffer_expand(&uart->rx, DEFAULT_BUFFER_SIZE);
+            SOL_INT_CHECK(err, < 0, true);
+        }
+
+        if (remaining > 0) {
```

Comment:
if remaining is still 0, you should go straight to the end, otherwise `r` will be undefined. You can also declare `ssize_t r = 0` above.